chain/consensus/mir/state_manager.go

```diff
@@ -273,146 +315,206 @@ func (sm *StateManager) ApplyTXs(txs []*requestpb.Request) error {
-	sm.memberships[nr+ConfigOffset] = sm.nextNewMembership
-
-	// Update current epoch number.
-	sm.currentEpoch = nr
+	err = sm.updateNextMembership(&valSet)
+	if err != nil {
+		return xerrors.Errorf("validator %v failed to update membership: %w", sm.ValidatorID, err)
+	}
 
-	// Garbage-collect previous membership and old voting data.
-	// Note that at initialization and after state transfer, these entries do not exist.
-	delete(sm.memberships, sm.currentEpoch-1)
-	delete(sm.reconfigurationVotes, sm.currentEpoch-1)
+	sm.nextConfigurationNumber = valSet.ConfigurationNumber
+	for n := range sm.reconfigurationVotes {
+		if n < sm.nextConfigurationNumber {
+			delete(sm.reconfigurationVotes, n)
+		}
+	}
 
-	return sm.nextNewMembership, nil
+	return nil
 }
 
-func (sm *StateManager) UpdateNextMembership(valSet *ValidatorSet) error {
-	_, mbs, err := validatorsMembership(valSet.GetValidators())
+func (sm *StateManager) updateNextMembership(valSet *validator.Set) error {
+	_, mbs, err := validator.Membership(valSet.GetValidators())
 	if err != nil {
 		return err
 	}
 	sm.nextNewMembership = mbs
+	log.With("validator", sm.ValidatorID).
+		Infof("updateNextMembership: current epoch %d, config number %d, next membership size: %d",
+			sm.currentEpoch, sm.nextConfigurationNumber, len(mbs))
 	return nil
 }
 
-// UpdateAndCheckVotes votes for the valSet and returns true if it has enough votes for this valSet.
-func (sm *StateManager) UpdateAndCheckVotes(valSet *ValidatorSet) (bool, error) {
-	h, err := valSet.Hash()
+// countVotes count votes for the validator set and returns true if we have got enough votes for this valSet.
+func (sm *StateManager) countVote(votingValidator t.NodeID, set *validator.Set) (bool, error) {
+	if set.ConfigurationNumber < sm.nextConfigurationNumber {
+		return false, xerrors.Errorf("validator %s sent outdated vote: received - %d, expected - %d",
+			votingValidator, set.ConfigurationNumber, sm.nextConfigurationNumber)
+	}
+
+	if _, found := sm.memberships[sm.currentEpoch][votingValidator]; !found {
+		return false, xerrors.Errorf("validator %s is not in the membership", votingValidator)
+	}
+
+	h, err := set.Hash()
 	if err != nil {
 		return false, err
 	}
-	_, ok := sm.reconfigurationVotes[sm.currentEpoch]
-	if !ok {
-		sm.reconfigurationVotes[sm.currentEpoch] = make(map[string]int)
+
+	if _, exist := sm.reconfigurationVotes[set.ConfigurationNumber]; !exist {
+		sm.reconfigurationVotes[set.ConfigurationNumber] = make(map[string][]t.NodeID)
 	}
-	sm.reconfigurationVotes[sm.currentEpoch][string(h)]++
-	votes := sm.reconfigurationVotes[sm.currentEpoch][string(h)]
+
+	// Prevent double voting.
+	for _, voted := range sm.reconfigurationVotes[set.ConfigurationNumber][string(h)] {
+		if voted == votingValidator {
+			return false, xerrors.Errorf("validator %s has been voted for configuration %d", votingValidator, set.ConfigurationNumber)
```

Comment:
has already voted, right? We are voting reconfiguration configs here, not validators.

Reply:
Correct. Fixed.